validphys2/src/validphys/commondataparser.py

```diff
@@ -1,18 +1,319 @@
+            necessary to create "gaps" that some cfactors or datasets might expect
+    """
+
+    repetition_flag: typing.Optional[typing.List[str]] = None
+    normalization: typing.Optional[dict] = None
+    shifts: typing.Optional[dict] = None
+
+    @classmethod
+    def parser(cls, meta: dict):
+        return parse_input(meta, cls)
+
+ValidApfelComb = Parser(ApfelComb.parser)
+
+@dataclass
+class TheoryMeta:
+    """Contains the necessary information to load the associated fktables"""
+
+    FK_tables: typing.List[list]
+    operation: ValidOperation
+    conversion_factor: float = 1.0 
+    comment: typing.Optional[str] = None
+    apfelcomb: typing.Optional[ValidApfelComb] = None
+    # The following options are transitional so that the yamldb can be used from the theory
+    appl: typing.Optional[bool] = False
+    target_dataset: typing.Optional[str] = None
+
+    def fktables_to_paths(self, grids_folder):
+        """Given a source for pineappl grids, constructs the lists of fktables
+        to be loaded"""
+        ret = []
+        for operand in self.FK_tables:
+            ret.append([grids_folder / f"{m}.{EXT}" for m in operand])
+        return ret
+    
+    @classmethod
+    def parser(cls, meta: dict):
+        """The yaml databases in the server use "operands" instead of "FK_tables" """
+        if "operands" in meta:
+            meta["FK_tables"] = meta.pop("operands")
+        return parse_input(meta, cls)
+
+
+@dataclass
+class KinematicsMeta:
+    """Contains all metadata for the kinematics of the dataset"""
+
+    file: ValidPath
+    variables: dict
+
+    @classmethod
+    def parser(cls, meta: dict):
+        parse_input(meta, cls)
+
+
+@dataclass
+class ReferenceMeta:
+    """Holds literature information for the dataset"""
+
+    url: str
+    version: int = 0
+    tables: typing.List[int] = field(default_factory=list)
+
+    @classmethod
+    def parser(cls, meta: dict):
+        return parse_input(meta, cls)
+
+
+@dataclass
+class Variant:
+    """Defines the keys of the CommonMetaData that can be overwritten"""
+
+    data_uncertainties: typing.List[ValidPath]
+
+
+# Define parsers for the more complicated structures
+ValidTheory = Parser(TheoryMeta.parser)
+ValidReference = Parser(ReferenceMeta.parser)
+ValidKinematics = Parser(KinematicsMeta.parser)
+
+
+@Parser
+def ValidVariants(variant_dict: dict) -> dict:
+    """Variants of a dataset are allowed to overwrite a subset of the keys of a dataset
+    (those defined in the Variant dataclass).
+    This wrapper class runs over the dictionary of variant and parses them into valid Variants
+    """
+    return {k: parse_input(v, Variant) for k, v in variant_dict.items()}
+
+
+@dataclass
+class CommonMetaData:
+    setname: str
+    ndata: int
+    observable: dict
+    kinematics: ValidKinematics
+    kinematic_coverage: dict
+    data_central: ValidPath
+    data_uncertainties: typing.List[ValidPath]
```

Comment:
I'd import things from `typing` than repeating it. Also `list[ValidPath]` works from Python 3.9.